internal/pkg/nats/options.go

```diff
@@ -0,0 +1,151 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//go:build include_nats_messaging
+
+package nats
+
+import (
+	"fmt"
+	"math/rand"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/edgexfoundry/go-mod-messaging/v2/internal/pkg"
+	"github.com/edgexfoundry/go-mod-messaging/v2/pkg/types"
+	"github.com/nats-io/nats.go"
+)
+
+// ClientConfig contains all the configurations for the MQTT client.
+type ClientConfig struct {
+	BrokerURL string
+	ClientOptions
+}
+
+// ConnectionOptions contains the connection configurations for the MQTT client.
+//
+// NOTE: The connection properties resides in its own struct in order to avoid the property being loaded in via
+//  reflection during the load process.
+type ConnectionOptions struct {
+	BrokerURL string
+}
+
+// ClientOptions contains the client options which are loaded via reflection
+type ClientOptions struct {
+	// Client Identifiers
+	Username string
+	Password string
+	ClientId string
+	// Connection information
+	Qos            int  // TODO: NATS core is QOS 0, Jetstream QOS 1 (or 2 with sync ack + msgid)
+	KeepAlive      int  // TODO: not needed
+	Retained       bool // TODO: not needed?  jetstream controls persistence
+	AutoReconnect  bool // TODO: not needed
+	CleanSession   bool // TODO: this probably maps to a start position somehow
+	ConnectTimeout int  // Seconds
+	pkg.TlsConfigurationOptions
+}
+
+// CreateClientConfiguration constructs a ClientConfig based on the provided MessageBusConfig.
+func CreateClientConfiguration(messageBusConfig types.MessageBusConfig) (ClientConfig, error) {
+	var brokerUrl string
+	if !messageBusConfig.PublishHost.IsHostInfoEmpty() {
+		brokerUrl = messageBusConfig.PublishHost.GetHostURL()
+	} else if !messageBusConfig.SubscribeHost.IsHostInfoEmpty() {
+		brokerUrl = messageBusConfig.SubscribeHost.GetHostURL()
+	} else {
+		return ClientConfig{}, fmt.Errorf("Specified empty broker info.")
+	}
+
+	_, err := url.Parse(brokerUrl)
+	if err != nil {
+		return ClientConfig{}, pkg.NewBrokerURLErr(fmt.Sprintf("Failed to parse broker: %v", err))
+	}
+
+	mqttClientOptions := CreateClientOptionsWithDefaults()
+	err = pkg.Load(messageBusConfig.Optional, &mqttClientOptions)
+	if err != nil {
+		return ClientConfig{}, err
+	}
+
+	tlsConfig := pkg.TlsConfigurationOptions{}
+	err = pkg.Load(messageBusConfig.Optional, &tlsConfig)
+	if err != nil {
+		return ClientConfig{}, err
+	}
+
+	mqttClientOptions.TlsConfigurationOptions = tlsConfig
+
+	return ClientConfig{
+		BrokerURL:     brokerUrl,
+		ClientOptions: mqttClientOptions,
+	}, nil
+}
+
+type connectType string
+
+const (
+	connectTypeSubscribe connectType = "subscribe"
+	connectTypePublish   connectType = "publish"
+	format                           = "format"
+	QueueGroup                       = "QueueGroup"
+)
+
+func (cc ClientConfig) connectOpt(ct connectType) []nats.Option {
+	connectTimeout := time.Second
+
+	if cc.ConnectTimeout != 0 {
```

Comment:
Yea might be able to do millisecond even it just needs to be positive to attempt I think.  Since ints=seconds for the MQTT stuff thats sort of the line of thought that got me to a second.